### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Isomorphisms

### PR DESCRIPTION
**Pull request by dupuisf**
This was stuck on an issue described in [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Another.20type.20class.20search.20issue).

---
- [x] depends on: #2967

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I think the root cause of the problem is https://github.com/leanprover/lean4/issue/2074

**Comment by dupuisf** [comment]
With `attribute [-instance] Ring.toNonAssocRing`, one can make some progress, but the file is *extremely* slow.

**Comment by Komyyy** [comment]
`synthInstance.etaExperiment` wasn't effective, maybe the cause is leanprover/lean4#2055.

**Comment by jcommelin** [comment]
Is this too slow to be merged?

**Comment by jcommelin** [comment]
I think the `maxHeartbeats` should come with a porting note. Ideally with a diagnosis of the problem in the form of a GH issue.

**Comment by urkud** [comment]
I'm going to try the `@` workaround (specify instances here and there) tonight.

**Comment by Komyyy** [comment]
I tried to build leanprover/lean4#2152 and build this file, but it wasn't effective.

File: Mathlib/LinearAlgebra/Isomorphisms.lean [hunk 1]
```diff
@@ -60,67 +62,92 @@ theorem quotKerEquivRange_apply_mk (x : M) :
...
       rintro ⟨x, hx⟩; rcases mem_sup.1 hx with ⟨y, hy, z, hz, rfl⟩
       use ⟨y, hy⟩; apply (Submodule.Quotient.eq _).2
       change y - (y + z) ∈ p'
-      rwa [sub_add_eq_sub_sub, sub_self, zero_sub, neg_mem_iff]⟩
+      rwa [sub_add_eq_sub_sub, sub_self, zero_sub, neg_mem_iff (H := p') (x := z)]⟩
 #align linear_map.quotient_inf_equiv_sup_quotient LinearMap.quotientInfEquivSupQuotient
 
-@[simp]
+set_option synthInstance.maxHeartbeats 200000 in
+-- @[simp] -- Porting note: `simp` affects `FunLike.coe`, so this theorem can't be a simp theorem
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Isomorphisms.lean:113]: What do you mean by `affects`?

File: Mathlib/LinearAlgebra/Isomorphisms.lean [hunk 2]
```diff
@@ -60,67 +62,92 @@ theorem quotKerEquivRange_apply_mk (x : M) :
...
-theorem quotKerEquivRange_symm_apply_image (x : M) (h : f x ∈ f.range) :
+theorem quotKerEquivRange_symm_apply_image (x : M) (h : f x ∈ LinearMap.range f) :
     f.quotKerEquivRange.symm ⟨f x, h⟩ = f.ker.mkQ x :=
   f.quotKerEquivRange.symm_apply_apply (f.ker.mkQ x)
 #align linear_map.quot_ker_equiv_range_symm_apply_image LinearMap.quotKerEquivRange_symm_apply_image
 
+/-
+Porting note: In these definitions and theorems, a match to `Submodule.instModuleSubtype
+  MemSubmoduleInstMembershipInstSetLikeSubmoduleInstAddCommMonoidSubtypeMemSubmoduleInstMembership
+  InstSetLikeSubmodule` which takes 0.1s happens a few hundreds times, so `maxHeartbeats` and
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Isomorphisms.lean:73]: I wonder if giving this a less outrageously long name would speed up hash lookups at all (marginally)...

Comment (Komyyy) [reply Mathlib/LinearAlgebra/Isomorphisms.lean:73]: #2967

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2967
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/LinearAlgebra/Isomorphisms.lean (hunk 1 above)
Comment (Komyyy) [reply Mathlib/LinearAlgebra/Isomorphisms.lean:113]:
```lean
@FunLike.coe
  ((↥p) ⧸ comap (Submodule.subtype p) (p ⊓ p') →ₗ[R] ↥(p ⊔ p') ⧸ comap (Submodule.subtype (p ⊔ p')) p')
  ((↥p) ⧸ comap (Submodule.subtype p) (p ⊓ p'))
  (fun a => ↥(p ⊔ p') ⧸ comap (Submodule.subtype (p ⊔ p')) p')
  _
  (quotientInfToSupQuotient p p')
===>
@FunLike.coe
  ((↥p) ⧸ comap (Submodule.subtype p) p ⊓ (↥p) ⧸ comap (Submodule.subtype p) p' →ₗ[R] ↥(p ⊔ p') ⧸ comap (Submodule.subtype (p ⊔ p')) p')
  ((↥p) ⧸ comap (Submodule.subtype p) p ⊓ (↥p) ⧸ comap (Submodule.subtype p) p')
  (fun a => ↥(p ⊔ p') ⧸ comap (Submodule.subtype (p ⊔ p')) p')
  _
  (quotientInfToSupQuotient p p')
```

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2967~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by mattrobball** [comment]
I outlined as much as I could but it is hard to outline `rfl`. If you could convince Lean of the FunLike instance on linear equivalences then I think everything else would go quickly.

**Comment by mattrobball** [comment]
I tried outlining `sup` on `Submodule`. That plus `etaExperiment` almost knocks the heartbeats down below 20000 on one other result (28000). It spends a lot of time time trying to chain together `CoeT` `CoeHCTC` `CoeOTC`... I confess to not understanding what all of these are meant to do.

File: Mathlib/LinearAlgebra/Isomorphisms.lean [hunk 3]
```diff
@@ -181,12 +192,16 @@ def quotientQuotientEquivQuotientAux (h : S ≤ T) : (M ⧸ S) ⧸ T.map S.mkQ 
...
+  submodule.quotient_quotient_equiv_quotient_aux_mk
+  Submodule.quotientQuotientEquivQuotientAux_mk
 
 -- @[simp] -- Porting note: simp can prove this
 theorem quotientQuotientEquivQuotientAux_mk_mk (x : M) :
     quotientQuotientEquivQuotientAux S T h (Quotient.mk (Quotient.mk x)) = Quotient.mk x := by simp
-#align submodule.quotient_quotient_equiv_quotient_aux_mk_mk Submodule.quotientQuotientEquivQuotientAux_mk_mk
+#align
+  submodule.quotient_quotient_equiv_quotient_aux_mk_mk
+  Submodule.quotientQuotientEquivQuotientAux_mk_mk
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Isomorphisms.lean]: Please don't line-wrap `#align`s

**Comment by mattrobball** [comment]
One possible solution: most of the timeout trouble comes from not being able to coerce a linear equivalence of (complicated) modules into a function. One can "cache" the TC synthesis in a result like 
```lean 
def asFun [AddCommMonoid N] [Module R N] (f : M ≃ₗ[R] N) : M → N := f 
```
as then wrap each `quotientInfEquivSupQuotient p p'` in `asFun` to avoid timing out.

File: Mathlib/LinearAlgebra/Isomorphisms.lean [hunk 4]
```diff
@@ -76,10 +76,14 @@ TODO: We hope these problem will be resolved by
   In that time, Please erase these `set_option`s.
 -/
 
+-- Porting note: breaking up original definition of quotientInfToSupQuotient to avoid timing out
 @[reducible]
-def aux' (p p' : Submodule R M) := (comap (p ⊔ p').subtype p').mkQ.comp (Submodule.ofLe le_sup_left)
+def subToSupQuotient (p p' : Submodule R M) :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Isomorphisms.lean:72]: Can you include the type annotation here?

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**